### PR TITLE
time projection operator

### PR DESCRIPTION
**Pull request by ChristianBeilschmidt**

### CODE REVIEW

**Comment by ChristianBeilschmidt** [comment]
[Vector-Time-Projection für Vektoren](https://trello.com/c/kNYH6GNq/619-vector-time-projection-f%C3%BCr-vektoren)

File: operators/src/processing/time_projection/mod.rs [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+    async fn initialize(
+        self: Box<Self>,
+        context: &dyn ExecutionContext,
+    ) -> Result<Box<dyn InitializedVectorOperator>> {
+        ensure!(self.params.step.step > 0, error::WindowSizeMustNotBeZero);
+
+        let source = self.sources.vector.initialize(context).await?;
+
+        debug!(
+            "Initializing `TemporalRasterAggregation` with {:?}.",
```
Comment (jdroenner) [inline operators/src/processing/time_projection/mod.rs:68]: TimeProjection

File: operators/src/processing/time_projection/mod.rs [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+            &self.params
+        );
+
+        let step_reference = self
+            .params
+            .step_reference
+            // use UTC 0 as default
+            .unwrap_or_else(|| TimeInstance::from_millis_unchecked(0));
```
Comment (jdroenner) [inline operators/src/processing/time_projection/mod.rs:76]: this needs to be very prominent in the docs i guess.

File: operators/src/processing/time_projection/mod.rs [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+                // TODO: change to [`try_collect_into_vec`](https://github.com/rayon-rs/rayon/issues/713) if available
+                .try_fold_with(Vec::new(), |mut acc, time_interval| {
+                    acc.push(time_interval?);
+                    Ok(acc)
+                })
+                .try_reduce_with(|a, b| Ok([a, b].concat()));
+
+        match time_intervals {
+            Some(Ok(time_intervals)) => feature_collection
+                .replace_time(&time_intervals)
```
Comment (jdroenner) [inline operators/src/processing/time_projection/mod.rs:193]: could we just add a new column as _the_ time column and keep the old information around? maybe make that a config option?

File: operators/src/processing/time_projection/mod.rs [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+use rayon::ThreadPool;
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, ResultExt, Snafu};
+
+/// Projection of time information in queries and data
+///
+/// This operator changes the temporal validity of the queried data.
+/// In order to query all valid data, it is necessary to change the query rectangle as well.
+///
+// TODO: extend for rasters
```
Comment (jdroenner) [inline operators/src/processing/time_projection/mod.rs:28]: i don't think we can implement this for raster data in a similar way. Raster data will always require aggregation.

File: operators/src/processing/time_projection/mod.rs [hunk 5]
```diff
@@ -0,0 +1,560 @@
...
+    };
+    use chrono::NaiveDate;
+    use geoengine_datatypes::{
+        collections::MultiPointCollection,
+        primitives::{BoundingBox2D, MultiPoint, SpatialResolution, TimeGranularity, TimeInterval},
+        util::test::TestDefault,
+    };
+
+    #[test]
+    fn test_expand_query_time_interval() {
```
Comment (jdroenner) [inline operators/src/processing/time_projection/mod.rs:249]: please add a test where the snap point is larger then the data time.

**Review by jdroenner — CHANGES_REQUESTED** [review]

File: operators/src/processing/time_projection/mod.rs (hunk 3 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/time_projection/mod.rs:193]: We have to wait for the new timestamp column type first.

File: operators/src/processing/time_projection/mod.rs (hunk 2 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/time_projection/mod.rs:73]: [20669f7](https://github.com/geo-engine/geoengine/pull/478/commits/20669f7e60b0ab37c250fdf3b1daaa807622f434)

File: operators/src/processing/time_projection/mod.rs (hunk 4 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/time_projection/mod.rs]: [4c652a4](https://github.com/geo-engine/geoengine/pull/478/commits/4c652a48661bb206769e7947d0640c4f1f33c567)

File: operators/src/processing/time_projection/mod.rs (hunk 5 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/time_projection/mod.rs:247]: [1f5a87b](https://github.com/geo-engine/geoengine/pull/478/commits/1f5a87be9fe672115f5aedefc8d96583bfd7f8bf)

File: operators/src/processing/time_projection/mod.rs (hunk 1 above)
Comment (ChristianBeilschmidt) [reply operators/src/processing/time_projection/mod.rs]: [20669f7](https://github.com/geo-engine/geoengine/pull/478/commits/20669f7e60b0ab37c250fdf3b1daaa807622f434)

**Review by jdroenner — APPROVED** [review]

**Comment by jdroenner** [comment]
bors merge

**Pull request merged**